File: Darts/src/burlton/dartzee/code/object/ColourWrapper.java
package burlton.dartzee.code.object;

import java.awt.Color;

public class ColourWrapper 
{
	private Color evenSingleColour;
	private Color evenDoubleColour;
	private Color evenTrebleColour;
	private Color oddSingleColour;
	private Color oddDoubleColour;
	private Color oddTrebleColour;
	private Color innerBullColour;
	private Color outerBullColour;
	private Color outerDartboardColour = Color.black;
	private Color missedBoardColour = null;
	
	//For wireframe mode
	private Color edgeColour = null;

	public ColourWrapper(Color singleColour)
	{
		this.evenSingleColour = singleColour;
		this.evenDoubleColour = singleColour;
		this.evenTrebleColour = singleColour;
		this.oddSingleColour = singleColour;
		this.oddDoubleColour = singleColour;
		this.oddTrebleColour = singleColour;
		this.innerBullColour = singleColour;
		this.outerBullColour = singleColour;
		this.outerDartboardColour = singleColour;
	}
	public ColourWrapper(Color evenSingleColour, Color evenDoubleColour,
			Color evenTrebleColour, Color oddSingleColour,
			Color oddDoubleColour, Color oddTrebleColour,
			Color innerBullColour, Color outerBullColour) 
	{
		this.evenSingleColour = evenSingleColour;
		this.evenDoubleColour = evenDoubleColour;
		this.evenTrebleColour = evenTrebleColour;
		this.oddSingleColour = oddSingleColour;
		this.oddDoubleColour = oddDoubleColour;
		this.oddTrebleColour = oddTrebleColour;
		this.innerBullColour = innerBullColour;
		this.outerBullColour = outerBullColour;
	}
	
	/**
	 * Helpers
	 */
	public Color getBullColour(int multiplier)
	{
		if (multiplier == 1)
		{
			return outerBullColour;
		}
		
		return innerBullColour;
	}
	
	public Color getColour(int multiplier, boolean even)
	{
		if (multiplier == 1)
		{
			return getSingleColour(even);
		}
		else if (multiplier == 2)
		{
			return getDoubleColour(even);
		}
		
		return getTrebleColour(even);
	}
	
	private Color getSingleColour(boolean even)
	{
		return even? evenSingleColour:oddSingleColour;
	}
	private Color getDoubleColour(boolean even)
	{
		return even? evenDoubleColour:oddDoubleColour;
	}
	private Color getTrebleColour(boolean even)
	{
		return even? evenTrebleColour:oddTrebleColour;
	}

	public Color getOuterDartboardColour()
	{
		return outerDartboardColour;
	}

	public void setOuterDartboardColour(Color outerDartboardColour)
	{
		this.outerDartboardColour = outerDartboardColour;
	}

	public Color getMissedBoardColour()
	{
		return missedBoardColour;
	}

	public void setMissedBoardColour(Color missedBoardColour)
	{
		this.missedBoardColour = missedBoardColour;
	}
	public Color getEdgeColour()
	{
		return edgeColour;
	}
	public void setEdgeColour(Color edgeColour)
	{
		this.edgeColour = edgeColour;
	}
}

File: src/main/kotlin/dartzee/object/ColourWrapper.kt
package dartzee.`object`

import dartzee.utils.DartsColour
import dartzee.utils.hmScoreToOrdinal
import java.awt.Color

val DEFAULT_COLOUR_WRAPPER =
    ColourWrapper(
        DartsColour.DARTBOARD_BLACK,
        Color.RED,
        Color.RED,
        Color.WHITE,
        Color.GREEN,
        Color.GREEN,
        Color.RED,
        Color.GREEN
    )

val WIREFRAME_COLOUR_WRAPPER =
    ColourWrapper(DartsColour.TRANSPARENT).also { it.edgeColour = Color.BLACK }

val GREY_COLOUR_WRAPPER =
    makeMonochromeWrapper(Color.GRAY.brighter(), Color.LIGHT_GRAY).also {
        it.outerDartboardColour = Color.GRAY
    }

fun makeMonochromeWrapper(dark: Color, light: Color) =
    ColourWrapper(dark, light, light, light, dark, dark, dark, light)

data class ColourWrapper(
    private val evenSingleColour: Color,
    private val evenDoubleColour: Color,
    private val evenTrebleColour: Color,
    private val oddSingleColour: Color,
    private val oddDoubleColour: Color,
    private val oddTrebleColour: Color,
    private val innerBullColour: Color,
    private val outerBullColour: Color,
    var outerDartboardColour: Color = Color.black
) {
    var missedBoardColour: Color = DartsColour.TRANSPARENT
    var edgeColour: Color? = null

    constructor(
        singleColour: Color
    ) : this(
        singleColour,
        singleColour,
        singleColour,
        singleColour,
        singleColour,
        singleColour,
        singleColour,
        singleColour,
        singleColour
    )

    /** Helpers */
    private fun getBullColour(multiplier: Int): Color {
        return when (multiplier) {
            1 -> outerBullColour
            else -> innerBullColour
        }
    }

    fun getColour(segment: DartboardSegment) = getColour(segment.getMultiplier(), segment.score)

    private fun getColour(multiplier: Int, score: Int): Color {
        if (score == 25) {
            return getBullColour(multiplier)
        }

        val even = hmScoreToOrdinal[Integer.valueOf(score)] ?: false
        return when (multiplier) {
            1 -> getSingleColour(even)
            2 -> getDoubleColour(even)
            3 -> getTrebleColour(even)
            else -> outerDartboardColour
        }
    }

    private fun getSingleColour(even: Boolean) = if (even) evenSingleColour else oddSingleColour

    private fun getDoubleColour(even: Boolean) = if (even) evenDoubleColour else oddDoubleColour

    private fun getTrebleColour(even: Boolean) = if (even) evenTrebleColour else oddTrebleColour
}
